classes/webservice/ApiNode.php

```diff
@@ -0,0 +1,404 @@
+     *                     'xlink_resource'    => [
+     *                     'resourceName'  => (string) Resource name, to url scheme: <i>'wsUrl/<strong>resourceName</strong>/<strong>subResourceName</strong>/<strong>object_id</strong>/<strong>value</strong>'</i>
+     *                     'subResourceName'  => (string) Sub-Resource name, used to compose output url.
+     *                     ]
+     *                     'object_id' => (string|int) Sub resource id. Used to compose output url and only if xlink_resource contains subResourceName.
+     *                     'getter'    => (string) Name of getter. If set, adds attribute <strong>notFilterable = true</strong>.
+     *                     'setter'    => (bool) Name of setter. If set to <i>false</i> and <strong>$schemaToDisplay</strong> is set to synopsis, adds attribute <strong>read_only = true</strong>.
+     *]
+     * </pre>
+     * @param string|null $schemaToDisplay 'blank' / 'synopsis'. Optional
+     *
+     * @return self
+     */
+    public function addField(array $field, ?string $schemaToDisplay = null): self
+    {
+        $newNode = self::value($field['sqlId']);
+
+        if (isset($field['encode'])) {
+            $newNode->addAttribute('encode', $field['encode']);
+        }
+
+        if (!empty($field['synopsis_details']) && $schemaToDisplay !== 'blank') {
+            foreach ($field['synopsis_details'] as $name => $detail) {
+                $newNode->addAttribute($name, is_array($detail) ? implode(' ', $detail) : $detail);
+            }
+        }
+
+        // display i18n fields
+        if (!empty($field['i18n'])) {
+            foreach (self::$languages as $language) {
+                $langAttributes = ['id' => $language];
+
+                if (isset($field['synopsis_details']) || (isset($field['value']) && is_array($field['value']))) {
+                    $langAttributes['xlink:href'] = WebserviceOutputBuilderCore::$wsUrl . 'languages/' . $language;
+                    if (isset($field['synopsis_details']) && $schemaToDisplay != 'blank') {
+                        $langAttributes['format'] = 'isUnsignedId';
+                    }
+                }
+
+                $newNode->setType(self::TYPE_LANGUAGE);
+                $newNode->addNode('language', $field['value'][$language] ?? '')
+                    ->setAttributes($langAttributes);
+            }
+        } else {
+            // display not i18n fields value
+            if (array_key_exists('xlink_resource', $field) && $schemaToDisplay != 'blank' && isset($field['value'])) {
+                if (!is_array($field['xlink_resource'])) {
+                    $xlink = WebserviceOutputBuilderCore::$wsUrl . $field['xlink_resource'] . '/' . $field['value'];
+                } else {
+                    $xlink = WebserviceOutputBuilderCore::$wsUrl . $field['xlink_resource']['resourceName'] . '/';
```

Comment:
As `WebserviceOutputBuilderCore::$wsUrl` is a public static attribute in a normal class, how do you ensure that it is set at this moment ?

The problem i see is that this `public function addField(array $field, ?string $schemaToDisplay = null): self` is almost a factory but the context it uses to build new objects is global.

I suspect we encounter some troubles if we try to test this code.

Reply:
Hello,
static property `$wsUrl` is actually replacement of class properties `WebserviceOutputBuilder::$wsUrl`, `WebserviceOutputJSON::$wsUrl` and `WebserviceOutputXML::$wsUrl`. Its being initialized frorm `WebserviceRequest->fetch(..)` method, which I believe is a starting point of each webservice request and then, during app flow, WebServiceOutputXX classes were being constructed using this `$wsUrl` string. Since `WebserviceOutputBuilder::$wsUrl` had been inited directly in constructor from `WebserviceRequest`, without any object-related logic, I made this variable static and declared on one place (`WebserviceOutputBuilder`) only, which made the code whole lot more readable (not mentioning slight performance gain).

It may seem like, but I really didnt rewrite the whole logic of PS WebService and actually didnt ask myself the question - how is ensured that this variable is always filled? I simply count that the one who wrote the `Webservice` was pretty sure :). Anyway, this variable is inited in `WebserviceOutputBuilder` constructor and looking at the constructing logic, guess I can be very sure that this variable is always inited.